arcgis-ios-sdk-samples/Geometry/Convex hull/ConvexHullViewController.swift

```diff
@@ -0,0 +1,121 @@
+    /// The graphics overlay for the convex hull.
+    let graphicsOverlay = AGSGraphicsOverlay()
+    
+    /// The graphic for the convex hull - comprised of either a point, a polyline or a polygon shape.
+    let convexHullGraphic = AGSGraphic(geometry: nil, symbol: nil, attributes: ["Type": "Hull"])
+    
+    /// List of geometry values (MapPoints in this case) that will be used by the AGSGeometryEngine.convexHull operation.
+    var inputPointArray: [AGSPoint] = []
+    
+    /// The bar button item that initiates the create convex hull operation.
+    @IBOutlet weak var creatConvexHullButtonItem: UIBarButtonItem!
+    
+    /// The bar button item that removes the convex hull as well as the MapPoints.
+    @IBOutlet weak var resetButtonItem: UIBarButtonItem!
+    
+    /// The map view managed by the view controller.
+    @IBOutlet weak var mapView: AGSMapView! {
+        didSet {
+            mapView.map = makeMap()
+            mapView.graphicsOverlays.add(graphicsOverlay)
+            mapView.touchDelegate = self
+        }
+    }
+    
+    /// Creates a map.
+    ///
+    /// - Returns: A new `AGSMap` object.
+    func makeMap() -> AGSMap {
+        let map = AGSMap(basemap: .topographic())
+        return map
+    }
+    
+    /// Called in response to the Create convex hull button being tapped.
+    @IBAction func createConvexHull() {
+        guard let convexHullGeometry = AGSGeometryEngine.convexHull(for: AGSMultipoint(points: inputPointArray)) else {
+            // Display the error as an alert if there is a problem with AGSGeometryEngine.convexHull operation.
+            let alertController = UIAlertController(title: nil, message: "Geometry Engine Failed!", preferredStyle: .alert)
+            alertController.addAction(UIAlertAction(title: "OK", style: .default))
+            present(alertController, animated: true)
+            return
+        }
+        // Change the symbol depending on the returned geometry type of the convex hull.
+        switch convexHullGeometry.geometryType {
+        case .point:
+            convexHullGraphic.symbol = ConvexHullViewController.markerSymbol
+        case .polyline:
+            convexHullGraphic.symbol = ConvexHullViewController.lineSymbol
+        case .polygon:
+            convexHullGraphic.symbol = ConvexHullViewController.fillSymbol
+        default:
+            convexHullGraphic.symbol = ConvexHullViewController.fillSymbol
+        }
+        convexHullGraphic.geometry = convexHullGeometry
+        creatConvexHullButtonItem.isEnabled = false
+    }
+    
+    /// Called in response to the Reset button being tapped.
+    @IBAction func reset() {
+        // Clear the existing points and graphics.
+        inputPointArray.removeAll()
+        graphicsOverlay.graphics.removeAllObjects()
+        // Reset the graphic for the convex hull and add it back to the to the graphics overlay.
+        convexHullGraphic.geometry = nil
+        graphicsOverlay.graphics.add(convexHullGraphic)
```

Comment:
I referred to the code for [`.NET`](https://github.com/Esri/arcgis-runtime-samples-dotnet/blob/53b8e9b604341a84a83add346a1d1c0f5b5374f8/src/WPF/ArcGISRuntime.WPF.Viewer/Samples/Geometry/ConvexHull/ConvexHull.xaml.cs#L121-L127) and their solution seems to be removing the previous convex hull graphic before adding the new one.

This solution has 2 main concerns
1. It will need to identify the graphic by its attribute name, which requires type casting
2. When I was testing with this method, when the amount of points grows (>16), it crashed for out of range. I didn't investigate that problem but rather had a stored property to avoid it...

Reply:
I'm sorry; I should have looked at how the sample works more closely. I didn't pick up on the fact that the existing points stayed when the convex hull is created. That being the case, I think it does make sense to have a property for the convex hull graphic. But instead of it always having a value, I would make it a `weak` optional and use it to keep track of the current graphic. `create()` can remove the current one (if there is one) before adding the new one and storing it as the current, and then it would get reset to `nil` when `reset()` removes all the graphics from the graphics overlay. That might be better than hunting through the `graphics` array for the convex hull graphic.

Reply:
FYI: The reason you were getting a crash is because `graphics` is an `NSMutableArray` (reference type), so you were mutating it while looping over the elements in it. Another downside is that you're looping over the whole array, even if the graphic you're looking for is in the middle or at the beginning. A better way would be to find the graphic first and then remove it:

```swift
if let graphic = graphicsOverlay.graphics.first(where:  { ($0 as! AGSGraphic).attributes["Type"] as! String == "Hull" }) {
    graphicsOverlay.graphics.remove(graphic)
}
```

Reply:
Yep that is what I thought. I almost never change an array during the loop, as it causes trouble. 

I used for loop with break to achieve the same as above, but it is much cleaner to use the first(where:) method! 👍 Anyway I've changed to the weak optional style.